# TABLEAU PROJECT

## Geospatial Business Intelligence (BI)


#### El objetivo del proyecto es determinar el punto geográfico donde recomendariamos abrir otra sede de nuestra empresa, teniendo en cuenta los siguientes datos:

   Debemos realizar un análisis geoespacial del dataset "companies"

   Datos sobre nuestra empresa:

   - Es una empresa de software de 50 empleados.

   - La empresa diseña videojuegos.

   - Puestos en nuestra compañía: 20 desarrolladores, 20 Diseñadores/creativos/UX/UI y 10 ejecutivos/managers

   Criterios para la localización de la nueva sede:

   - Debe haber ingenieros de software alrededor

   - Debe haber un buen ratio de grandes compañías vs startups

   - Asegurate de que las compañias cercanas cubren los intereses de tu equipo

   - Evita las viejas empresas, rodeate de empresas nuevas

        

1. Importamos las librerias necesarias para trabajar:

In [3]:
import pymongo
MongoClient = pymongo.MongoClient
import time
client = MongoClient()
import pandas as pd
import re
db = client.companies

In [4]:
new_companies = db.companies.find({'founded_year': { '$gte': 2008} })
len(list(new_companies))

3308

2. Para seleccionar empresas que se dediquen a actividades similares a la nuestra, buscamos todos los "category_code" de las empresas de la BBDD:

In [5]:
cursor2 = db.companies.find({'category_code':{'$exists': 'true'}})
category_codes = list(map(lambda r:r['category_code'], list(cursor2)))

category_code_clean = []
for i in category_codes:
    if i not in category_code_clean:
        category_code_clean.append(i)
    else:
        pass
print(category_code_clean)

['enterprise', 'web', 'software', 'news', 'network_hosting', 'games_video', 'music', 'social', 'mobile', 'search', 'advertising', 'messaging', 'security', 'photo_video', 'finance', 'ecommerce', 'hardware', 'travel', 'public_relations', 'other', 'real_estate', 'semiconductor', 'analytics', 'health', 'legal', 'sports', 'biotech', 'cleantech', 'education', 'consulting', 'transportation', None, 'hospitality', 'fashion', 'nonprofit', 'nanotech', 'automotive', 'design', 'manufacturing', 'government', 'local', 'medical']


3. Vamos a seleccionar las empresas que se dedican a : 'web', 'software', 'network_hosting', 'games_video', 'ecommerce', 'hardware', 'analytics':

In [6]:
category_companies = db.companies.find({'$and':
                                       [{'founded_year': { '$gte': 2008} }, 
                                         { '$or': 
                                          [{'category_code': 'web'}, 
                                           {'category_code':'software'}, 
                                           {'category_code':'network_hosting'},
                                           {'category_code': 'games_video'},
                                           {'category_code':'ecommerce'},
                                           {'category_code':'hardware'}, 
                                           {'category_code':'network_hosting'},
                                           {'category_code': 'games_video'},
                                           {'category_code':'ecommerce'},
                                           {'category_code':'hardware'}]}]})
len(list(category_companies))

1971

4. Nos quedamos con las empresas que no tengan año de Deadpool:

In [7]:
category_companies_no_dp = db.companies.find({'$and':
                                       [{'founded_year': { '$gte': 2008} }, 
                                         { '$or': 
                                          [{'category_code': 'web'}, 
                                           {'category_code':'software'}, 
                                           {'}category_code':'network_hosting'},
                                           {'category_code': 'games_video'},
                                           {'category_code':'ecommerce'},
                                           {'category_code':'hardware'}, 
                                           {'category_code':'network_hosting'},
                                           {'category_code': 'games_video'},
                                           {'category_code':'ecommerce'},
                                           {'category_code':'hardware'}]},
                                       {'deadpooled_year': { '$type': 10 }}]})
len(list(category_companies_no_dp))

1824

5. Analizamos los datos del campo "total_money_raised":

In [10]:
cursor3 = db.companies.find({'total_money_raised':{'$exists': 'true'}})
total_money_raised = list(map(lambda r:r['total_money_raised'], list(cursor3)))

total_money_raised_list = []
for i in total_money_raised:
    if i not in total_money_raised_list:
        total_money_raised_list.append(i)
    else:
        pass
print(max(total_money_raised_list))
print(min(total_money_raised_list))

€9M
$0


Eliminamos las que total_money_raised = '$0':

In [11]:
category_companies_no_dp_act = db.companies.find({'$and':
                                       [{'founded_year': { '$gte': 2008} }, 
                                         { '$or': 
                                          [{'category_code': 'web'}, 
                                           {'category_code':'software'}, 
                                           {'category_code':'network_hosting'},
                                           {'category_code': 'games_video'},
                                           {'category_code':'ecommerce'},
                                           {'category_code':'hardware'}, 
                                           {'category_code':'network_hosting'},
                                           {'category_code': 'games_video'},
                                           {'category_code':'ecommerce'},
                                           {'category_code':'hardware'}]},
                                       {'deadpooled_year': { '$type': 10 }},
                                       {'total_money_raised':{'$nin': ['$0']}}]})
x = list(category_companies_no_dp_act)
len(x)

245

6. Del listado de las 245 empresas, creamos la Dataframe: 

In [13]:
df = pd.DataFrame(x)
df.head()

_id  \
0  52cdef7c4bab8bd675297e30   
1  52cdef7c4bab8bd675297f76   
2  52cdef7c4bab8bd675297fec   
3  52cdef7c4bab8bd67529801f   
4  52cdef7c4bab8bd67529803d   

                                         acquisition  \
0                                               None   
1                                               None   
2                                               None   
3                                               None   
4  {'price_amount': None, 'price_currency_code': ...   

                                        acquisitions alias_list  \
0                                                 []       None   
1                                                 []              
2                                                 []              
3                                                 []              
4  [{'price_amount': None, 'price_currency_code':...              

                     blog_feed_url                     blog_url category_code  \
0                                       http://goingon.com/blog      software   
1                                                                 games_video   
2     http://blog.muktar.com/feed/       http://blog.muktar.com           web   
3                                         http://corp.wamba.com           web   
4  http://www.uptake.com/blog/feed  http://www.uptake.com/blog/           web   

                                        competitions  \
0  [{'competitor': {'name': 'Ning by Glam Media',...   
1                                                 []   
2  [{'competitor': {'name': 'Kluster', 'permalink...   
3  [{'competitor': {'name': 'Vostu', 'permalink':...   
4  [{'competitor': {'name': 'TripAdvisor', 'perma...   

                     created_at  \
0  Fri Jul 20 04:55:26 UTC 2007   
1  Thu Aug 30 11:29:22 UTC 2007   
2  Sat Sep 08 08:24:15 UTC 2007   
3  Fri Sep 28 09:37:07 UTC 2007   
4  Thu Oct 11 12:07:18 UTC 2007   

                                  crunchbase_url  \
0      http://www.crunchbase.com/company/goingon   
1  http://www.crunchbase.com/company/crunchyroll   
2        http://www.crunchbase.com/company/fixya   
3        http://www.crunchbase.com/company/wamba   
4       http://www.crunchbase.com/company/uptake   

                         ...                            phone_number  \
0                        ...                            415-200-4606   
1                        ...                                           
2                        ...                                           
3                        ...                                           
4                        ...                          (650) 329-9888   

                                            products  \
0  [{'name': 'GoingOn', 'permalink': 'goingon'}, ...   
1  [{'name': 'crunchyroll.com', 'permalink': 'cru...   
2          [{'name': 'Fixya', 'permalink': 'fixya'}]   
3          [{'name': 'Wamba', 'permalink': 'wamba'}]   
4         [{'name': 'Uptake', 'permalink': 'kango'}]   

                                       providerships  \
0                                                 []   
1                                                 []   
2  [{'title': 'Accounting', 'is_past': False, 'pr...   
3  [{'title': 'IT Services Provider', 'is_past': ...   
4  [{'title': '', 'is_past': False, 'provider': {...   

                                       relationships screenshots  \
0  [{'is_past': False, 'title': 'CEO', 'person': ...          []   
1  [{'is_past': False, 'title': 'beneficial owner...          []   
2  [{'is_past': False, 'title': 'Founder & CEO', ...          []   
3  [{'is_past': True, 'title': 'Founder', 'person...          []   
4  [{'is_past': False, 'title': 'VP of Revenue', ...          []   

                                            tag_list total_money_raised  \
0  socialnetworking, enterprise, social-media, so...              $8.5M   
1  crunchyroll, anime, manga, video, gaming, musi...              $4.8M   
2  fixya

6. Eliminamos las filas cuya columna Offices no tiene info:

In [16]:
df['l'] = df.offices.apply(len)

7. Creamos dos columnas ('lat' y 'long'), con los datos que estaban incluidos en la columna 'office':

In [17]:
df1 = df.loc[df['l'] > 0]
df1.reset_index(inplace=True)
df1.offices[0]
df1.offices[0][0]['longitude'] # df.offices -> el primer registro, la primera office, latitude

df1['lat'] = df1.apply(lambda row: row['offices'][0]['latitude'], axis=1)
df1['long'] = df1.apply(lambda row: row['offices'][0]['longitude'], axis=1)

/home/laura/.local/lib/python3.6/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/home/laura/.local/lib/python3.6/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


8. Eliminamos las filas que no tienen datos de 'lat' y 'ĺong':

In [19]:
df1['lat'].isnull().sum()

df2 = df1.dropna(subset=['lat'])
df2.reset_index(inplace=True)
df2.columns

Index(['level_0', 'index', '_id', 'acquisition', 'acquisitions', 'alias_list',
       'blog_feed_url', 'blog_url', 'category_code', 'competitions',
       'created_at', 'crunchbase_url', 'deadpooled_day', 'deadpooled_month',
       'deadpooled_url', 'deadpooled_year', 'description', 'email_address',
       'external_links', 'founded_day', 'founded_month', 'founded_year',
       'funding_rounds', 'homepage_url', 'image', 'investments', 'ipo',
       'milestones', 'name', 'number_of_employees', 'offices', 'overview',
       'partners', 'permalink', 'phone_number', 'products', 'providerships',
       'relationships', 'screenshots', 'tag_list', 'total_money_raised',
       'twitter_username', 'updated_at', 'video_embeds', 'l', 'lat', 'long'],
      dtype='object')

8. Unimos latitud y longitud en una columna, que llamaremos "coord":

In [20]:
df2['coord'] = df2[['long', 'lat']].apply(list, axis=1)
df2.coord

/home/laura/.local/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


0        [-122.392142, 37.782263]
1        [-122.393229, 37.781265]
2        [-122.323895, 37.566879]
3        [-122.161674, 37.443058]
4         [-73.985433, 40.740231]
5        [-118.445243, 34.047312]
6         [-85.717393, 38.257035]
7        [-122.161523, 37.446823]
8         [-74.370313, 40.219214]
9           [5.291266, 52.132633]
10        [-73.985281, 40.723652]
11       [-122.211435, 37.503147]
12       [-122.214897, 37.772323]
13       [-118.243425, 34.052187]
14        [-118.495981, 34.01052]
15       [-122.155678, 37.446382]
16       [-118.422495, 34.153588]
17       [-122.419204, 37.775196]
18         [-73.980306, 40.66509]
19     [-122.0253784, 36.9751032]
20     [-122.4013624, 37.7893215]
21          [-71.05682, 42.35888]
22       [-119.306607, 37.269175]
23          [-0.208979, 51.51819]
24             [34.7595, 32.0554]
25          [0.257568, 51.125693]
26        [-73.985506, 40.757929]
27         [-71.24762, 42.275263]
28       [-113.515769, 53.544711]
29       [-111

In [21]:
def toGeoJSON(array):
    return {
       "type":"Point",
       "coordinates":array}
df2['position'] = df2.apply(lambda r:toGeoJSON(r['coord']),axis=1)
df2.position


/home/laura/.local/lib/python3.6/site-packages/ipykernel_launcher.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app


0      {'type': 'Point', 'coordinates': [-122.392142,...
1      {'type': 'Point', 'coordinates': [-122.393229,...
2      {'type': 'Point', 'coordinates': [-122.323895,...
3      {'type': 'Point', 'coordinates': [-122.161674,...
4      {'type': 'Point', 'coordinates': [-73.985433, ...
5      {'type': 'Point', 'coordinates': [-118.445243,...
6      {'type': 'Point', 'coordinates': [-85.717393, ...
7      {'type': 'Point', 'coordinates': [-122.161523,...
8      {'type': 'Point', 'coordinates': [-74.370313, ...
9      {'type': 'Point', 'coordinates': [5.291266, 52...
10     {'type': 'Point', 'coordinates': [-73.985281, ...
11     {'type': 'Point', 'coordinates': [-122.211435,...
12     {'type': 'Point', 'coordinates': [-122.214897,...
13     {'type': 'Point', 'coordinates': [-118.243425,...
14     {'type': 'Point', 'coordinates': [-118.495981,...
15     {'type': 'Point', 'coordinates': [-122.155678,...
16     {'type': 'Point', 'coordinates': [-118.422495,...
17     {'type': 'Point', 'coord

8. Seleccionamos las columnas que nos interesan:

In [22]:
df3=df2[['name','category_code','founded_year','number_of_employees', 'total_money_raised', 'position']]
df3.head()

name category_code  founded_year  number_of_employees  \
0      GoingOn      software          2008                 40.0   
1  Crunchyroll   games_video          2008                 50.0   
2        Fixya           web          2013                 30.0   
3       Uptake           web          2008                  NaN   
4    BeliefNet           web          2008                  NaN   

  total_money_raised                                           position  
0              $8.5M  {'type': 'Point', 'coordinates': [-122.392142,...  
1              $4.8M  {'type': 'Point', 'coordinates': [-122.393229,...  
2                $8M  {'type': 'Point', 'coordinates': [-122.323895,...  
3             $16.8M  {'type': 'Point', 'coordinates': [-122.161674,...  
4                $7M  {'type': 'Point', 'coordinates': [-73.985433, ...

9. Pasamos el data a .JSON

In [23]:
df3.to_json("companies_with_coords.json", orient="records", lines=True)

10. En MongoDB ya hemos creado un index geographical en nuestro .JOSN:

In [24]:
cursor4= db.companiesjson.find()
companies_found = []
for i in cursor4:
    cursor5 = db.companiesjson.find({
              "position": {
                "$near": {
                  "$geometry": {
                    "type": "Point",
                    "coordinates": [i['position']['coordinates'][0], i['position']['coordinates'][1]]
      },
      "$minDistance": 0,
      "$maxDistance":10000
    }
  }
})
    companies_found.append(cursor5.count())

print(max(companies_found))
    
#import pandas as pd
#df4 = pd.DataFrame(list(cursor4))
#df4.head()

/home/laura/.local/lib/python3.6/site-packages/ipykernel_launcher.py:16: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  app.launch_new_instance()


25


In [26]:
cursor4= db.companiesjson.find()
companies_found = []
for i in cursor4:
    cursor5 = db.companiesjson.find({
              "position": {
                "$near": {
                  "$geometry": {
                    "type": "Point",
                    "coordinates": [i['position']['coordinates'][0], i['position']['coordinates'][1]]
      },
      "$minDistance": 0,
      "$maxDistance":10000
    }
  }
})
    if cursor5.count() == 25:
        companies_found.append(i)
    
import pandas as pd
df4 = pd.DataFrame(list(companies_found))
df4

/home/laura/.local/lib/python3.6/site-packages/ipykernel_launcher.py:16: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  app.launch_new_instance()


_id category_code  founded_year               name  \
0   5c6af01fadf5549f750d37a8           web          2008          BeliefNet   
1   5c6af01fadf5549f750d37af           web          2008            Kluster   
2   5c6af01fadf5549f750d37c5           web          2008              Bitly   
3   5c6af02fadf5549f750d37f9           web          2010              Yipit   
4   5c6af02fadf5549f750d37fd           web          2008          Modelinia   
5   5c6af02fadf5549f750d381b   games_video          2008       Crispy Gamer   
6   5c6af02fadf5549f750d3823     ecommerce          2008       Proper Cloth   
7   5c6af02fadf5549f750d3827      software          2009  GameChanger Media   
8   5c6af02fadf5549f750d3828     ecommerce          2008           IntheGlo   
9   5c6af02fadf5549f750d3831   games_video          2008              Moogi   
10  5c6af02fadf5549f750d3843           web          2009         Catchafire   
11  5c6af02fadf5549f750d384c           web          2008        StyleCaster   
12  5c6af02fadf5549f750d3854           web          2009      ChallengePost   

    number_of_employees                                           position  \
0                   NaN  {'type': 'Point', 'coordinates': [-73.985433, ...   
1                   NaN  {'type': 'Point', 'coordinates': [-73.985281, ...   
2                   NaN  {'type': 'Point', 'coordinates': [-74.00721, 4...   
3                  23.0  {'type': 'Point', 'coordinates': [-73.987764, ...   
4                   NaN  {'type': 'Point', 'coordinates': [-73.990024, ...   
5                   NaN  {'type': 'Point', 'coordinates': [-73.985273, ...   
6                   1.0  {'type': 'Point', 'coordinates': [-74.008767, ...   
7                  20.0  {'type': 'Point', 'coordinates': [-74.0136605,...   
8                   5.0  {'type': 'Point', 'coordinates': [-73.987764, ...   
9                   NaN  {'type': 'Point', 'coordinates': [-74.0094471,...   
10                  NaN  {'type': 'Point', 'coordinates': [-73.983171, ...   
11                 60.0  {'type': 'Point', 'coordinates': [-73.9900984,...   
12                 17.0  {'type': 'Point', 'coordinates': [-74.00717, 4...   

   total_money_raised  
0                 $7M  
1                 $1M  
2              $31.4M  
3              $7.55M  
4                 $5M  
5              $8.25M  
6               $100k  
7              $3.64M  
8                $30k  
9               $100k  
10              $400k  
11                $5M  
12              $4.6M

In [51]:
df4.position[0]['coordinates'][0]   #el primer registro, la primera coordinates, (longitude)

df4['long'] = df4.apply(lambda row: row['position']['coordinates'][0], axis=1)
df4['lat'] = df4.apply(lambda row: row['position']['coordinates'][1], axis=1)

df4.head()

_id category_code  founded_year       name  \
0  5c6af01fadf5549f750d37a8           web          2008  BeliefNet   
1  5c6af01fadf5549f750d37af           web          2008    Kluster   
2  5c6af01fadf5549f750d37c5           web          2008      Bitly   
3  5c6af02fadf5549f750d37f9           web          2010      Yipit   
4  5c6af02fadf5549f750d37fd           web          2008  Modelinia   

   number_of_employees                                           position  \
0                  NaN  {'type': 'Point', 'coordinates': [-73.985433, ...   
1                  NaN  {'type': 'Point', 'coordinates': [-73.985281, ...   
2                  NaN  {'type': 'Point', 'coordinates': [-74.00721, 4...   
3                 23.0  {'type': 'Point', 'coordinates': [-73.987764, ...   
4                  NaN  {'type': 'Point', 'coordinates': [-73.990024, ...   

  total_money_raised        lat       long  
0                $7M  40.740231 -73.985433  
1                $1M  40.723652 -73.985281  
2             $31.4M  40.740340 -74.007210  
3             $7.55M  40.744618 -73.987764  
4                $5M  40.739867 -73.990024

11. Nos quedamos solo con las columnas que nos interesa mostrar en Tableau:

In [58]:
df4.drop(['_id', 'position'], inplace=True, axis=1)

KeyError: "['_id' 'position'] not found in axis"

12. Guardamos nuestro resultado como .JSON: 

In [60]:
df4.to_json('visualize_companies', orient="records", lines=True)